### PR TITLE
upgrade evm to cancun 

### PR DESCRIPTION
**Pull request by wenlinlee**

### CODE REVIEW

File: bcos-executor/src/executor/TransactionExecutor.cpp [hunk 1]
```diff
@@ -411,6 +414,20 @@ BlockContext::Ptr TransactionExecutor::createBlockContext(
 
     return context;
 }
+VMSchedule const& TransactionExecutor::getVMSchedule(uint32_t currentVersion)
+{
+    ledger::Features features;
+    task::syncWait(features.readFromStorage(*m_backendStorage, m_lastCommittedBlockNumber + 1));
+    if (features.get(ledger::Features::Flag::feature_evm_cancun))
+    {
+        return FiscoBcosScheduleCanCun;
```
Comment (bxq2011hust) [inline bcos-executor/src/executor/TransactionExecutor.cpp:423]:
```suggestion
        return FiscoBcosScheduleCancun;
```

File: bcos-executor/src/executive/TransactionExecutive.h [hunk 2]
```diff
@@ -61,14 +62,18 @@ class TransactionExecutive : public std::enable_shared_from_this<TransactionExec
...
         m_seq(seq),
         m_gasInjector(gasInjector),
         m_storageWrapperObj(m_blockContext.storage(), m_recoder),
-        m_storageWrapper(&m_storageWrapperObj)
+        m_transientStorageWrapperObj(m_transientStateStorage, m_transientRecoder),
+        m_storageWrapper(&m_storageWrapperObj),
+        m_transientStorageWrapper(&m_transientStorageWrapperObj)
     {
         m_recoder = std::make_shared<storage::Recoder>();
+        m_transientRecoder = std::make_shared<storage::Recoder>();
```
Comment (bxq2011hust) [inline bcos-executor/src/executive/TransactionExecutive.h:76]: construct m_transientRecoder before it is used

**Review by bxq2011hust — CHANGES_REQUESTED** [review]

File: bcos-executor/src/vm/Precompiled.cpp [hunk 3]
```diff
@@ -522,6 +567,317 @@ pair<bool, bytes> ecRecover(bytesConstRef _in)
...
+//         }
+//         state->total_len_delivered = true;
+//     } else {
+//         memset(chunk, 0x00, space_in_chunk);
+//     }
+//
+//     return true;
+// }
+//
+// static inline ALWAYS_INLINE void sha_256_implementation(uint32_t h[8], const void* input, size_t
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/Precompiled.cpp:646]: add a new file to put 0xa implement

File: bcos-executor/src/vm/VMInstance.cpp [hunk 4]
```diff
@@ -60,8 +59,9 @@ Result VMInstance::execute(HostContext& _hostContext, evmc_message* _msg)
         return Result(m_instance->execute(m_instance, _hostContext.interface, &_hostContext,
             m_revision, _msg, m_code.data(), m_code.size()));
     }
-    auto state = std::make_unique<evmone::advanced::AdvancedExecutionState>(
-        *_msg, m_revision, *_hostContext.interface, &_hostContext, m_code);
+    bytes_view data{};
+    auto state = std::make_unique<evmone::ExecutionState>(
+        *_msg, m_revision, *_hostContext.interface, &_hostContext, m_code, data);
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/VMInstance.cpp:64]:
```suggestion
        *_msg, m_revision, *_hostContext.interface, &_hostContext, m_code, {});
```

File: bcos-executor/src/vm/HostContext.h [hunk 5]
```diff
@@ -64,10 +64,12 @@ class HostContext : public evmc_host_context
 
     /// Read storage location.
     evmc_bytes32 store(const evmc_bytes32* key);
+    evmc_bytes32 storeTransient(const evmc_bytes32* key);
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/HostContext.h:67]:
```suggestion
    evmc_bytes32 getTransient(const evmc_bytes32* key);
```

File: bcos-executor/src/vm/EVMHostInterface.cpp [hunk 6]
```diff
@@ -85,6 +97,21 @@ evmc_storage_status setStorage(evmc_host_context* context,
...
 }
 
+void setTransientStorage(evmc_host_context* context, [[maybe_unused]] const evmc_address* addr,
+    const evmc_bytes32* key, const evmc_bytes32* value)
+{
+    auto& hostContext = static_cast<HostContext&>(*context);
+
+    assert(fromEvmC(*addr) == boost::algorithm::unhex(std::string(hostContext.myAddress())));
+    // TODO: use evmc_storage_status 5-8
+    if (value == 0)  // TODO: Should use 32 bytes 0
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/EVMHostInterface.cpp:107]: delete this, the code is useless

File: bcos-executor/src/vm/Precompiled.cpp (hunk 3 above)
Comment (wenlinlee) [reply bcos-executor/src/vm/Precompiled.cpp]: next pr

File: bcos-executor/src/vm/HostContext.h [hunk 7]
```diff
@@ -64,10 +64,12 @@ class HostContext : public evmc_host_context
 
     /// Read storage location.
     evmc_bytes32 store(const evmc_bytes32* key);
+    evmc_bytes32 getTransient(const evmc_bytes32* key);
 
     /// Write a value in storage.
     // void setStore(const u256& _n, const u256& _v);
     void setStore(const evmc_bytes32* key, const evmc_bytes32* value);
+    void setTransientStore(const evmc_bytes32* key, const evmc_bytes32* value);
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/HostContext.h:72]:
```suggestion
    void setTransient(const evmc_bytes32* key, const evmc_bytes32* value);
```

File: bcos-executor/src/vm/HostContext.cpp [hunk 8]
```diff
@@ -570,6 +587,16 @@ void HostContext::setStore(const evmc_bytes32* key, const evmc_bytes32* value)
     m_executive->storage().setRow(m_tableName, keyView, std::move(entry));
 }
 
+void HostContext::setTransientStore(const evmc_bytes32* key, const evmc_bytes32* value)
```
Comment (bxq2011hust) [inline bcos-executor/src/vm/HostContext.cpp:590]:
```suggestion
void HostContext::setTransient(const evmc_bytes32* key, const evmc_bytes32* value)
```

**Review by bxq2011hust — DISMISSED** [review]

**Review by morebtcg — DISMISSED** [review]

**Review by morebtcg — DISMISSED** [review]

**Review by kyonRay — DISMISSED** [review]

**Review by morebtcg — DISMISSED** [review]

File: bcos-executor/src/executive/BlockContext.cpp [hunk 9]
```diff
@@ -66,16 +66,24 @@ BlockContext::BlockContext(std::shared_ptr<storage::StateStorageInterface> stora
...
 
     task::syncWait(m_features.readFromStorage(*m_storage, m_blockNumber));
+    if (m_features.get(ledger::Features::Flag::feature_evm_cancun))
+    {
+        m_schedule = FiscoBcosScheduleCancun;
+    }
+    if (blockVersion >= (uint32_t)bcos::protocol::BlockVersion::V3_2_VERSION)
+    {
+        m_schedule = FiscoBcosScheduleV320;
+    }
```
Comment (JimmyShi22) [inline bcos-executor/src/executive/BlockContext.cpp:76]:
```suggestion
    if (m_features.get(ledger::Features::Flag::feature_evm_cancun))
    {
        m_schedule = FiscoBcosScheduleCancun;
    }
    else if (blockVersion >= (uint32_t)bcos::protocol::BlockVersion::V3_2_VERSION)
    {
        m_schedule = FiscoBcosScheduleV320;
    } else 
    {
        m_schedule = FiscoBcosSchedule;
    }
```

File: bcos-executor/src/executive/BlockContext.cpp [hunk 10]
```diff
@@ -102,6 +110,14 @@ BlockContext::BlockContext(std::shared_ptr<storage::StateStorageInterface> stora
...
         }
     }
+    if (_current->version() >= (uint32_t)bcos::protocol::BlockVersion::V3_2_VERSION)
+    {
+        m_schedule = FiscoBcosScheduleV320;
+    }
+    if (m_features.get(ledger::Features::Flag::feature_evm_cancun))
+    {
+        m_schedule = FiscoBcosScheduleCancun;
+    }
```
Comment (JimmyShi22) [inline bcos-executor/src/executive/BlockContext.cpp:120]: 封装一个函数，复用之前的逻辑

File: bcos-executor/src/Common.h [hunk 11]
```diff
@@ -187,6 +188,15 @@ static const VMSchedule FiscoBcosScheduleV320 = [] {
     return schedule;
 }();
 
+static const VMSchedule FiscoBcosScheduleCancun = [] {
+    VMSchedule schedule = VMSchedule();
+    schedule.enablePairs = false;
+    schedule.enableCanCun = true;
```
[the remark this answers is not in the document]
Comment (wenlinlee) [reply bcos-executor/src/Common.h:194]: 这里我理解标志位是代表使用哪个版本，用于判断toversion函数里面返回的版本号，enableCanCun位true就代表version是cancun，enablePairs为true就代表所用版本为巴黎

File: bcos-executor/src/executor/TransactionExecutor.cpp [hunk 12]
```diff
@@ -411,6 +414,20 @@ BlockContext::Ptr TransactionExecutor::createBlockContext(
 
     return context;
 }
+VMSchedule const& TransactionExecutor::getVMSchedule(uint32_t currentVersion)
+{
+    ledger::Features features;
+    task::syncWait(features.readFromStorage(*m_backendStorage, m_lastCommittedBlockNumber + 1));
```
[the remark this answers is not in the document]
Comment (wenlinlee) [reply bcos-executor/src/executor/TransactionExecutor.cpp]: 这里已去掉

File: bcos-executor/src/executive/TransactionExecutive.cpp [hunk 13]
```diff
@@ -1875,3 +1879,44 @@ std::string TransactionExecutive::getContractTableName(
 
     return std::string(USER_APPS_PREFIX).append(formatAddress);
 }
+
+std::shared_ptr<storage::StateStorageInterface> TransactionExecutive::getTransientStateStorage(
```
[the remark this answers is not in the document]
Comment (wenlinlee) [reply bcos-executor/src/executive/TransactionExecutive.cpp:1883]: 把此函数放至BlockContext，在TransactionExecutive无法被调用

File: bcos-executor/src/executive/BlockContext.cpp [hunk 14]
```diff
@@ -45,16 +45,16 @@ using namespace std;
...
     storage::StorageInterface::Ptr backendStorage)
   : m_blockNumber(blockNumber),
     m_blockHash(blockHash),
     m_timeStamp(timestamp),
     m_blockVersion(blockVersion),
-    m_schedule(_schedule),
     m_isWasm(_isWasm),
     m_isAuthCheck(_isAuthCheck),
     m_storage(std::move(storage)),
+    m_transientStorageMap(std::make_shared<transientStorageMap>(10)),
```
Comment (bxq2011hust) [inline bcos-executor/src/executive/BlockContext.cpp:57]: why not use std::map?
the transientStorage's lifetime is tx, every tx has its transientStorage. I think we don't need to consider parallel

**Review by bxq2011hust — APPROVED** [review]

**Review by morebtcg — APPROVED** [review]

**Review by kyonRay — APPROVED** [review]

**Pull request merged**